zapcore/write_syncer.go

```diff
@@ -75,6 +77,117 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	sync.Mutex
+	stop chan struct{}
+
+	bufferWriter *bufio.Writer
+	ticker       *time.Ticker
+}
+
+const (
+	// _defaultBufferSize specifies the default size used by Buffer.
+	_defaultBufferSize = 256 * 1024 // 256 kB
+
+	// _defaultFlushInterval specifies the default flush interval for
+	// Buffer.
+	_defaultFlushInterval = 30 * time.Second
+)
+
+// Buffer wraps a WriteSyncer to buffer its output. The returned WriteSyncer
+// flushes its output as the buffer fills up, or at the provided interval,
+// whichever comes first.
+//
+// Call the returned function to finish using the WriteSyncer and flush
+// remaining bytes.
+//
+//   func main() {
+//     // ...
+//     ws, closeWS := zapcore.Buffer(ws, 0, 0)
+//     defer closeWS()
+//     // ...
+//   }
+//
+// The buffer size defaults to 256 kB if set to zero.
+// The flush interval defaults to 30 seconds if set to zero.
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (_ WriteSyncer, close func() error) {
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws, func() error { return nil }
+	}
+
+	if bufferSize == 0 {
+		bufferSize = _defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = _defaultFlushInterval
+	}
+
+	ticker := time.NewTicker(flushInterval)
+
+	bws := &bufferWriterSyncer{
+		stop:         make(chan struct{}),
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+		ticker:       ticker,
+	}
+
+	go bws.flushLoop()
+
+	return bws, bws.close
+}
+
+// Write writes log data into buffer syncer directly, multiple Write calls will be batched,
+// and log data will be flushed to disk when the buffer is full or periodically.
+func (s *bufferWriterSyncer) Write(bs []byte) (int, error) {
+	// bufio is not goroutine safe, so add lock writer here
+	s.Lock()
+	defer s.Unlock()
+
+	// there are some logic internal for bufio.Writer here:
+	// 1. when the buffer is enough, data would not be flushed.
+	// 2. when the buffer is not enough, data would be flushed as soon as the buffer fills up.
+	// this would lead to log spliting, which is not acceptable for log collector
+	// so we need to flush bufferWriter before writing the data into bufferWriter
```

Comment:
Let me try and simplify and clarify this comment,
```suggestion
	// To avoid partial writes from being flushed, we manually flush the existing buffer if:
	// * The current write doesn't fit into the buffer fully, and
	// * The buffer is not empty (since bufio will not split large writes when the buffer is empty)
```